test/distributed/org/apache/cassandra/distributed/action/GossipHelper.java

```diff
@@ -0,0 +1,431 @@
+                                                                 String partitionerStr, String initialTokenStr)
+    {
+        return instance.appliesOnInstance((String partitionerString, String tokenString) -> {
+            IPartitioner partitioner = FBUtilities.newPartitioner(partitionerString);
+            Token token = partitioner.getTokenFactory().fromString(tokenString);
+
+            VersionedValue versionedValue = supplier.apply(partitioner, Collections.singleton(token));
+            return new VersionedApplicationState(applicationState.ordinal(), versionedValue.value, versionedValue.version);
+        }).apply(partitionerStr, initialTokenStr);
+    }
+
+    public static VersionedApplicationState versionedToken(IInvokableInstance instance, ApplicationState applicationState, SerializableBiFunction<IPartitioner, Collection<Token>, VersionedValue> supplier)
+    {
+        return unsafeVersionedToken(instance, applicationState, supplier, instance.config().getString("partitioner"), instance.config().getString("initial_token"));
+    }
+
+    public void removeFromRing(IInstance peer)
+    {
+        InetAddressAndPort endpoint = toCassandraInetAddressAndPort(peer.broadcastAddress());
+
+        TokenMetadata tokenMetadata = StorageService.instance.getTokenMetadata();
+        tokenMetadata.removeEndpoint(endpoint);
+        Collection<Token> tokens = tokenMetadata.getTokens(endpoint);
+        if (!tokens.isEmpty())
+            tokenMetadata.removeBootstrapTokens(tokens);
+
+        Gossiper.runInGossipStageBlocking(() -> Gossiper.instance.unsafeAnulEndpoint(endpoint));
+
+        SystemKeyspace.removeEndpoint(endpoint);
+        PendingRangeCalculatorService.instance.update();
+        PendingRangeCalculatorService.instance.blockUntilFinished();
+    }
+
+    public static void changeGossipState(Stream<IInvokableInstance> targets, Instance peer, List<VersionedApplicationState> newState)
+    {
+        targets.forEach((instance) -> {
+            changeGossipState(instance, peer, newState);
+        });
+    }
+
+    /**
+     * Changes gossip state of the `peer` on `target`
+     */
+    public static void changeGossipState(IInvokableInstance target, IInstance peer, List<VersionedApplicationState> newState)
+    {
+        int version = peer.isShutdown() ? MessagingService.current_version : peer.getMessagingVersion();
+
+        target.appliesOnInstance((InetSocketAddress addr, UUID hostId, Integer messagingVersion) -> {
+            InetAddressAndPort endpoint = toCassandraInetAddressAndPort(addr);
+            StorageService storageService = StorageService.instance;
+
+            Gossiper.runInGossipStageBlocking(() -> {
+                EndpointState state = Gossiper.instance.getEndpointStateForEndpoint(endpoint);
+                if (state == null)
+                {
+                    Gossiper.instance.initializeNodeUnsafe(endpoint, hostId, 1);
```

Comment:
this sets `ApplicationState.NET_VERSION` to target's version rather than the peer's, should this be updated to take `version`?

Reply:
Since i've switched to use net_version below, I think it should be ok here.

Reply:
for safety can we remove the state?  This is a public function with variable input, so nothing forces the input to include NET_VERSION

Reply:
the change was to add it based off the peer state, but if its replaced; cool.

+1